### PR TITLE
[ADDED] MQTT SparkplugB Aware support

### PR DESCRIPTION
**Pull request by levb**
# SpakplugB-aware support. 

## TODO:
- [x] Add spec comments
- [x] Add go test
- [x] Resolve `[tck-id-conformance-mqtt-aware-nbirth-mqtt-retain]`, `[tck-id-conformance-mqtt-aware-dbirth-mqtt-retain]` issues
- [ ] Decide on a config option to enable SparkplugB awareness, get benchmark comparisons

## SPEC

[PDF](https://sparkplug.eclipse.org/specification/version/3.0/documents/sparkplug-specification-3.0.0.pdf)

#### 12.65. Sparkplug Compliant MQTT Server
- [x] [tck-id-conformance-mqtt-qos0] A Sparkplug conformant MQTT Server MUST support publish and
subscribe on QoS 0
- [x] [tck-id-conformance-mqtt-qos1] A Sparkplug conformant MQTT Server MUST support publish and
subscribe on QoS 1
- [x] [tck-id-conformance-mqtt-will-messages] A Sparkplug conformant MQTT Server MUST support all
133
aspects of Will Messages including use of the retain flag and QoS 1
- [x] [tck-id-conformance-mqtt-retained] A Sparkplug conformant MQTT Server MUST support all
aspects of the retain flag

#### 12.66. Sparkplug Aware MQTT Server
- [x] [tck-id-conformance-mqtt-aware-basic] A Sparkplug Aware MQTT Server MUST support all aspects
of a Sparkplug Compliant MQTT Server
- [x] [tck-id-conformance-mqtt-aware-store] A Sparkplug Aware MQTT Server MUST store NBIRTH and
DBIRTH messages as they pass through the MQTT Server
- [x] [tck-id-conformance-mqtt-aware-nbirth-mqtt-topic] A Sparkplug Aware MQTT Server MUST make
NBIRTH messages available on a topic of the form:
$sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id
- [x] [tck-id-conformance-mqtt-aware-nbirth-mqtt-retain] A Sparkplug Aware MQTT Server MUST make
NBIRTH messages available on the topic:
$sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id with the MQTT retain flag set to
true
- [x] [tck-id-conformance-mqtt-aware-dbirth-mqtt-topic] A Sparkplug Aware MQTT Server MUST make
DBIRTH messages available on a topic of the form:
$sparkplug/certificates/namespace/group_id/DBIRTH/edge_node_id/device_id
- [x] [tck-id-conformance-mqtt-aware-dbirth-mqtt-retain] A Sparkplug Aware MQTT Server MUST make
DBIRTH messages available on the topic:
$sparkplug/certificates/namespace/group_id/DBIRTH/edge_node_id/device_id with the MQTT retain
flag set to true
- [x] [tck-id-conformance-mqtt-aware-ndeath-timestamp] A Sparkplug Aware MQTT Server MAY replace
the timestmap of NDEATH messages. If it does, it MUST set the timestamp to the UTC time at which
it attempts to deliver the NDEATH to subscribed clients

## ISSUES:
- [x] `[tck-id-conformance-mqtt-aware-nbirth-mqtt-retain]`, `[tck-id-conformance-mqtt-aware-dbirth-mqtt-retain]` - not 100% clear from the spec if the messages should be published (only!) with the retained flag on, or should also be retained for future subscriptions (presently implemented). `[tck-id-conformance-mqtt-aware-store]` seems to suggest that they should indeed be stored as retained.
- [x] `[tck-id-conformance-mqtt-aware-ndeath-timestamp]` Not presently implemented. The spec says MAY, but TCK seems to require it. Adding re-writing the timestamp involves modifying protobuf payload, which would likely necessitate importing lots of go dependencies. Passed TCK (by faking the DEATH timestamp in the test node and device).
image

### CODE REVIEW

File: server/mqtt.go
Comment (levb) [inline server/mqtt.go:4041]: Not related: seems like we have been over-allocating here.

File: server/mqtt.go
Comment (levb) [inline server/mqtt.go:4317]: @kozlovic This will send new notification messages on `$sparkplug...` subject. They need to be delivered as QoS0 messages, so thankfully sending a NATS message is sufficient. I figured I can just leverage `jsa` for it, but it uses the system account, not 100% if this is proper. 

```
[tck-id-conformance-mqtt-aware-nbirth-mqtt-retain] A Sparkplug Aware MQTT Server MUST make
NBIRTH messages available on the topic:
$sparkplug/certificates/namespace/group_id/NBIRTH/edge_node_id with the MQTT retain flag set to
true
```

**Comment by levb** [comment]
@kozlovic @derekcollison This is ready for review, sans the timestamp/protobuf issue.

File: server/mqtt.go
Comment (kozlovic) [inline server/mqtt.go:4636]: Why do we have to convert? Meaning do we expect messages to be from NATS? I guess it does not hurt because if they are already "MQTT topic-like" tokens, then should not be any conversion needed.

File: server/mqtt.go
Comment (kozlovic) [reply server/mqtt.go:4326]: No, the asm.jsa.sendMsg() will send using an internal client on the account the session was created for, so in other words, will use the account of the incoming published message.

File: server/mqtt.go
Comment (kozlovic) [inline server/mqtt.go:4815]: This will enqueue an MQTT message to the matching subscriber, but this is not going to "produce" a retained message, that is, this is not going to be stored as a retained message. Does that make sense?

File: server/mqtt_test.go
Comment (kozlovic) [inline server/mqtt_test.go:7597]: `truncated`

File: server/mqtt_test.go
Comment (kozlovic) [inline server/mqtt_test.go:7555]: I don't see that you are using NATS subscriptions to monitor in this test.

File: server/mqtt_test.go
Comment (kozlovic) [inline server/mqtt_test.go:7617]: `make`, and put the next line `there.` on this comment line. The code after that goes farther anyway.

File: server/mqtt.go
Comment (kozlovic) [inline server/mqtt.go:4284]: For a sparkb message, the pp.flags will indicate that the message is not retained, correct? Should it be switched to retained since we are storing it at such?

**Review by kozlovic — CHANGES_REQUESTED** [review]
Some changes to make but also wondering if we should have a server flag to run as a SparkplugB compliance/support mode or do all this by default? I mean producing messages on the spark subjects in all cases will have a cost, if anything for the server to process the produced messages and possibly not finding matching subscriptions.

File: server/mqtt.go
Comment (levb) [reply server/mqtt.go:4271]: Right, see [tck-id-conformance-mqtt-aware-store] and [tck-id-conformance-mqtt-aware-nbirth-mqtt-retain]

This is one of the 2 "open" issues - I am not 100% sure if "store" means we should store it as retained.

File: server/mqtt.go
Comment (levb) [reply server/mqtt.go:4638]: I just refactored this in a separate branch where I worked on overwriting the DEATH timestamp. Once I implement all other feedback, I'll see if it makes sense to merge it here, or submit as a separate PR.

File: server/mqtt.go
Comment (levb) [reply server/mqtt.go:4817]: Yes, we stored it as a retained in `mqttHandlePubRetain`. Here we need to set the retained flag as per spec; the original BIRTH message must not have retain set (also per spec).

Comment (levb) [reply server/mqtt.go:4817]: Oh I see what you mean - we stored the retained for the BIRTH itself, but not for the BIRTH certificate. I'll review the spec once more.

File: server/mqtt_test.go
Comment (levb) [reply server/mqtt_test.go]: Oops, sry was leftover from the first implementation.

File: server/mqtt.go
Comment (levb) [reply server/mqtt.go:4638]: See https://github.com/nats-io/nats-server/pull/5251; 1/5 I prefer to keep it a separate PR, it will probably have a number of comments of its own. But if it makes it easier for you to review - feel free to merge into this branch.

**Comment by levb** [comment]
>  if we should have a server flag to run as a SparkplugB compliance/support mode or do all this by default?

Obviously, would be easy to add, but not sure if it's really worth it.

> I mean producing messages on the spark subjects in all cases will have a cost, if anything for the server to process the produced messages and possibly not finding matching subscriptions.

We apply the prefix filtering to (all) MQTT incoming messages, but only produce certificates for BIRTHs (nodes and devices). I am 0/5. @derekcollison?

**Review by kozlovic — COMMENTED** [review]
Reviewed the other PR. Since they intersect, let's finish the other one too and maybe have a final review after that.

File: server/mqtt.go
Comment (levb) [reply server/mqtt.go]: I think this reflects the spec. On a "normal" MQTT broker the BIRTH messages are not retained at all; that's the baseline. Awareness adds publishing (and retaining) the birth certificates. It makes sense that they are retained - they are supposed to contain the full set of (all) last known metrics values.

**Review by kozlovic — APPROVED** [review]
LGTM

**Comment by derekcollison** [comment]
This ok to merge? And if so, squash or normal?

**Comment by levb** [comment]
@derekcollison we should decide if we want to have the config option to hide the SparkplugB support away. Unfortunately, I've been trying to gather an adequate benchmark comparisons and my batches (`--count 10`) either keep timing out or the results end up too noisy. Can you please hold off another day or so as I try to get them to run in a more controlled environment than my dev machine?

As of now, I haven't seen a statistically significant difference, but I don't have entire clean batches

**Comment by derekcollison** [comment]
Keep me posted.

**Comment by levb** [comment]
@derekcollison @kozlovic I couldn't get the noise down much, I failed to produce differences beyond the noise. Most recent batch I ran is pasted below. 0/5 we merge as is.

```
goos: darwin
goarch: arm64
pkg: github.com/nats-io/nats-server/v2/server
                                  │ lev-main.out │             lev-spB.out             │
                                  │    sec/op    │    sec/op     vs base               │
MQTT_QoS0_Pub_______0b_Payload-12   103.9n ±  2%   105.2n ±  1%       ~ (p=0.309 n=10)
MQTT_QoS0_Pub_______8b_Payload-12   110.9n ±  3%   108.0n ±  2%       ~ (p=0.066 n=10)
MQTT_QoS0_Pub______32b_Payload-12   115.7n ±  2%   115.5n ±  1%       ~ (p=0.839 n=10)
MQTT_QoS0_Pub_____128b_Payload-12   163.2n ±  2%   166.9n ±  1%  +2.33% (p=0.001 n=10)
MQTT_QoS0_Pub_____256b_Payload-12   201.7n ±  1%   204.6n ±  1%  +1.46% (p=0.012 n=10)
MQTT_QoS0_Pub_______1K_Payload-12   428.7n ±  6%   444.6n ±  2%  +3.71% (p=0.004 n=10)
MQTT_QoS0_PubSub1___0b_Payload-12   367.7n ±  1%   368.0n ±  2%       ~ (p=0.984 n=10)
MQTT_QoS0_PubSub1___8b_Payload-12   382.2n ±  2%   379.6n ±  1%       ~ (p=0.382 n=10)
MQTT_QoS0_PubSub1__32b_Payload-12   372.4n ±  2%   377.9n ±  2%  +1.50% (p=0.019 n=10)
MQTT_QoS0_PubSub1_128b_Payload-12   480.3n ±  1%   469.1n ±  1%  -2.34% (p=0.000 n=10)
MQTT_QoS0_PubSub1_256b_Payload-12   559.6n ±  3%   547.9n ±  1%  -2.08% (p=0.000 n=10)
MQTT_QoS0_PubSub1___1K_Payload-12   867.2n ±  2%   866.4n ±  1%       ~ (p=0.927 n=10)
MQTT_QoS0_PubSub2___0b_Payload-12   544.2n ±  3%   547.8n ±  1%       ~ (p=0.671 n=10)
MQTT_QoS0_PubSub2___8b_Payload-12   552.8n ±  2%   566.7n ±  1%  +2.51% (p=0.000 n=10)
MQTT_QoS0_PubSub2__32b_Payload-12   575.2n ±  2%   575.5n ±  1%       ~ (p=0.753 n=10)
MQTT_QoS0_PubSub2_128b_Payload-12   699.1n ±  3%   695.1n ±  1%       ~ (p=0.190 n=10)
MQTT_QoS0_PubSub2_256b_Payload-12   776.9n ±  7%   796.6n ±  1%  +2.53% (p=0.023 n=10)
MQTT_QoS0_PubSub2___1K_Payload-12   1.167µ ±  5%   1.185µ ±  2%       ~ (p=0.279 n=10)
MQTT_QoS1_Pub_______0b_Payload-12   5.908µ ±  3%   5.787µ ±  1%       ~ (p=0.060 n=10)
MQTT_QoS1_Pub_______8b_Payload-12   6.057µ ±  4%   5.753µ ±  3%  -5.02% (p=0.005 n=10)
MQTT_QoS1_Pub______32b_Payload-12   6.075µ ±  4%   5.689µ ± 24%  -6.35% (p=0.042 n=10)
MQTT_QoS1_Pub_____128b_Payload-12   6.061µ ±  2%   5.774µ ± 27%       ~ (p=0.137 n=10)
MQTT_QoS1_Pub_____256b_Payload-12   6.108µ ±  2%   5.937µ ± 11%  -2.79% (p=0.035 n=10)
MQTT_QoS1_Pub_______1K_Payload-12   6.780µ ±  3%   6.734µ ±  8%       ~ (p=0.853 n=10)
MQTT_QoS1_PubSub1___0b_Payload-12   16.20µ ±  6%   16.12µ ±  2%       ~ (p=0.289 n=10)
MQTT_QoS1_PubSub1___8b_Payload-12   15.73µ ±  3%   15.92µ ±  2%       ~ (p=0.516 n=10)
MQTT_QoS1_PubSub1__32b_Payload-12   15.68µ ±  2%   15.56µ ±  3%       ~ (p=0.105 n=10)
MQTT_QoS1_PubSub1_128b_Payload-12   16.10µ ±  2%   15.82µ ±  2%       ~ (p=0.089 n=10)
MQTT_QoS1_PubSub1_256b_Payload-12   16.37µ ±  3%   16.80µ ±  3%       ~ (p=0.075 n=10)
MQTT_QoS1_PubSub1___1K_Payload-12   18.23µ ± 14%   18.44µ ±  4%       ~ (p=0.353 n=10)
MQTT_QoS1_PubSub2___0b_Payload-12   18.46µ ±  6%   19.10µ ±  5%       ~ (p=0.075 n=10)
MQTT_QoS1_PubSub2___8b_Payload-12   18.57µ ±  7%   18.36µ ±  8%       ~ (p=0.631 n=10)
MQTT_QoS1_PubSub2__32b_Payload-12   19.04µ ±  8%   18.89µ ±  5%       ~ (p=0.739 n=10)
MQTT_QoS1_PubSub2_128b_Payload-12   18.55µ ± 10%   19.45µ ±  2%  +4.88% (p=0.035 n=10)
MQTT_QoS1_PubSub2_256b_Payload-12   18.82µ ± 30%   20.27µ ±  6%       ~ (p=0.165 n=10)
MQTT_QoS1_PubSub2___1K_Payload-12   24.29µ ± 12%   22.23µ ±  5%       ~ (p=0.165 n=10)
geomean                             2.177µ         2.173µ        -0.19%

                                  │ lev-main.out  │             lev-spB.out              │
                                  │      B/s      │      B/s       vs base               │
MQTT_QoS0_Pub_______0b_Payload-12   45.88Mi ±  2%   45.33Mi ±  1%       ~ (p=0.315 n=10)
MQTT_QoS0_Pub_______8b_Payload-12   111.8Mi ±  3%   114.8Mi ±  2%       ~ (p=0.063 n=10)
MQTT_QoS0_Pub______32b_Payload-12   305.0Mi ±  2%   305.6Mi ±  1%       ~ (p=0.853 n=10)
MQTT_QoS0_Pub_____128b_Payload-12   783.4Mi ±  2%   765.4Mi ±  1%  -2.29% (p=0.002 n=10)
MQTT_QoS0_Pub_____256b_Payload-12   1.210Gi ±  1%   1.193Gi ±  1%  -1.44% (p=0.012 n=10)
MQTT_QoS0_Pub_______1K_Payload-12   2.238Gi ±  6%   2.158Gi ±  2%  -3.57% (p=0.004 n=10)
MQTT_QoS0_PubSub1___0b_Payload-12   25.94Mi ±  1%   25.92Mi ±  2%       ~ (p=0.984 n=10)
MQTT_QoS0_PubSub1___8b_Payload-12   64.87Mi ±  2%   65.32Mi ±  1%       ~ (p=0.382 n=10)
MQTT_QoS0_PubSub1__32b_Payload-12   189.5Mi ±  2%   186.7Mi ±  2%  -1.49% (p=0.019 n=10)
MQTT_QoS0_PubSub1_128b_Payload-12   532.1Mi ±  1%   544.9Mi ±  1%  +2.41% (p=0.000 n=10)
MQTT_QoS0_PubSub1_256b_Payload-12   892.9Mi ±  3%   912.0Mi ±  1%  +2.13% (p=0.000 n=10)
MQTT_QoS0_PubSub1___1K_Payload-12   2.212Gi ±  2%   2.214Gi ±  1%       ~ (p=0.912 n=10)
MQTT_QoS0_PubSub2___0b_Payload-12   26.28Mi ±  3%   26.11Mi ±  1%       ~ (p=0.671 n=10)
MQTT_QoS0_PubSub2___8b_Payload-12   67.28Mi ±  2%   65.63Mi ±  1%  -2.45% (p=0.000 n=10)
MQTT_QoS0_PubSub2__32b_Payload-12   184.0Mi ±  2%   183.9Mi ±  1%       ~ (p=0.796 n=10)
MQTT_QoS0_PubSub2_128b_Payload-12   548.4Mi ±  3%   551.6Mi ±  1%       ~ (p=0.190 n=10)
MQTT_QoS0_PubSub2_256b_Payload-12   964.8Mi ±  7%   941.0Mi ±  1%  -2.47% (p=0.023 n=10)
MQTT_QoS0_PubSub2___1K_Payload-12   2.467Gi ±  5%   2.430Gi ±  2%       ~ (p=0.280 n=10)
MQTT_QoS1_Pub_______0b_Payload-12   1.779Mi ±  3%   1.812Mi ±  2%  +1.88% (p=0.028 n=10)
MQTT_QoS1_Pub_______8b_Payload-12   2.990Mi ±  4%   3.152Mi ±  3%  +5.42% (p=0.005 n=10)
MQTT_QoS1_Pub______32b_Payload-12   6.752Mi ±  5%   7.210Mi ± 20%  +6.78% (p=0.041 n=10)
MQTT_QoS1_Pub_____128b_Payload-12   22.03Mi ±  2%   23.13Mi ± 21%       ~ (p=0.138 n=10)
MQTT_QoS1_Pub_____256b_Payload-12   41.85Mi ±  2%   43.05Mi ± 10%  +2.86% (p=0.034 n=10)
MQTT_QoS1_Pub_______1K_Payload-12   145.7Mi ±  3%   146.7Mi ±  8%       ~ (p=0.811 n=10)
MQTT_QoS1_PubSub1___0b_Payload-12   1.297Mi ±  6%   1.302Mi ±  2%       ~ (p=0.359 n=10)
MQTT_QoS1_PubSub1___8b_Payload-12   2.303Mi ±  3%   2.279Mi ±  3%       ~ (p=0.560 n=10)
MQTT_QoS1_PubSub1__32b_Payload-12   5.231Mi ±  2%   5.274Mi ±  3%       ~ (p=0.101 n=10)
MQTT_QoS1_PubSub1_128b_Payload-12   16.58Mi ±  2%   16.88Mi ±  2%       ~ (p=0.089 n=10)
MQTT_QoS1_PubSub1_256b_Payload-12   31.23Mi ±  3%   30.42Mi ±  3%       ~ (p=0.075 n=10)
MQTT_QoS1_PubSub1___1K_Payload-12   108.4Mi ± 12%   107.2Mi ±  4%       ~ (p=0.353 n=10)
MQTT_QoS1_PubSub2___0b_Payload-12   1.702Mi ±  6%   1.645Mi ±  6%       ~ (p=0.060 n=10)
MQTT_QoS1_PubSub2___8b_Payload-12   2.928Mi ±  7%   2.966Mi ±  7%       ~ (p=0.590 n=10)
MQTT_QoS1_PubSub2__32b_Payload-12   6.461Mi ±  9%   6.514Mi ±  5%       ~ (p=0.753 n=10)
MQTT_QoS1_PubSub2_128b_Payload-12   21.60Mi ±  9%   20.59Mi ±  2%  -4.66% (p=0.035 n=10)
MQTT_QoS1_PubSub2_256b_Payload-12   40.74Mi ± 23%   37.83Mi ±  6%       ~ (p=0.165 n=10)
MQTT_QoS1_PubSub2___1K_Payload-12   122.2Mi ± 13%   133.4Mi ±  5%       ~ (p=0.165 n=10)
geomean                             59.50Mi         59.61Mi        +0.19%

                                  │ lev-main.out │              lev-spB.out              │
                                  │     B/op     │     B/op      vs base                 │
MQTT_QoS0_Pub_______0b_Payload-12     52.00 ± 0%     52.00 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_______8b_Payload-12     77.00 ± 0%     77.00 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub______32b_Payload-12     120.0 ± 0%     120.0 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_____128b_Payload-12     332.0 ± 0%     332.0 ± 0%       ~ (p=0.474 n=10)
MQTT_QoS0_Pub_____256b_Payload-12     605.0 ± 0%     605.0 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_______1K_Payload-12   2.286Ki ± 0%   2.286Ki ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1___0b_Payload-12     293.0 ± 0%     293.0 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1___8b_Payload-12     319.0 ± 0%     319.0 ± 0%       ~ (p=1.000 n=10)
MQTT_QoS0_PubSub1__32b_Payload-12     365.0 ± 0%     365.0 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1_128b_Payload-12     589.0 ± 0%     589.0 ± 0%       ~ (p=0.054 n=10)
MQTT_QoS0_PubSub1_256b_Payload-12     878.5 ± 0%     878.0 ± 0%       ~ (p=0.207 n=10)
MQTT_QoS0_PubSub1___1K_Payload-12   2.545Ki ± 0%   2.544Ki ± 0%  -0.04% (p=0.008 n=10)
MQTT_QoS0_PubSub2___0b_Payload-12     405.0 ± 0%     405.0 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2___8b_Payload-12     432.0 ± 0%     433.0 ± 0%       ~ (p=0.370 n=10)
MQTT_QoS0_PubSub2__32b_Payload-12     482.0 ± 0%     482.0 ± 0%       ~ (p=1.000 n=10)
MQTT_QoS0_PubSub2_128b_Payload-12     718.5 ± 0%     718.0 ± 0%       ~ (p=0.350 n=10)
MQTT_QoS0_PubSub2_256b_Payload-12    1023.0 ± 0%    1024.0 ± 0%       ~ (p=0.120 n=10)
MQTT_QoS0_PubSub2___1K_Payload-12   2.680Ki ± 0%   2.678Ki ± 0%  -0.07% (p=0.003 n=10)
geomean                               479.1          479.1       +0.01%
¹ all samples are equal

                                  │ lev-main.out │             lev-spB.out             │
                                  │  allocs/op   │ allocs/op   vs base                 │
MQTT_QoS0_Pub_______0b_Payload-12     3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_______8b_Payload-12     3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub______32b_Payload-12     3.000 ± 0%   3.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_____128b_Payload-12     4.000 ± 0%   4.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_____256b_Payload-12     4.000 ± 0%   4.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_Pub_______1K_Payload-12     4.000 ± 0%   4.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1___0b_Payload-12     6.000 ± 0%   6.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1___8b_Payload-12     6.000 ± 0%   6.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1__32b_Payload-12     6.000 ± 0%   6.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1_128b_Payload-12     7.000 ± 0%   7.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1_256b_Payload-12     7.000 ± 0%   7.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub1___1K_Payload-12     7.000 ± 0%   7.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2___0b_Payload-12     8.000 ± 0%   8.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2___8b_Payload-12     8.000 ± 0%   8.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2__32b_Payload-12     8.000 ± 0%   8.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2_128b_Payload-12     9.000 ± 0%   9.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2_256b_Payload-12     9.000 ± 0%   9.000 ± 0%       ~ (p=1.000 n=10) ¹
MQTT_QoS0_PubSub2___1K_Payload-12     9.000 ± 0%   9.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                               5.754        5.754       +0.00%
¹ all samples are equal
```

**Comment by levb** [comment]
(and @derekcollison please SQUASH on merge. Squashing the PR would make the comments/history out of place)

**Pull request merged**
